File: state_models/TireState.py
from utilities import MF52
import setups
import numpy as np

class TireState():
    # TODO: setups thing here
    def __init__(self, params: setups.Goose):
        self._MF52 = MF52()
        self.params = params

        #Slip angle
        self.alpha = 0

        #Longitudinal Slip
        self.kappa = 0

        #Inclination angle
        self.epsilon = 0
        
        #Normal Load
        self.Fz = 0

        self.Fx = 0 #N
        self.Fy = 0 #N
    
    '''Returns the unscaled longitudinal force in the tire frame'''
    def eval_Fx(self):
        fx_range = []
        for sl in np.linspace(0, 0.2, 100):
            fx_range.append(self._MF52.Fx(self.Fz, sl, self.epsilon) * self.params.friction_scaling_x)
        self.Fx = max(fx_range)
    
    '''Returns the unscaled lateral force in the tire frame'''
    def eval_Fy(self):
        self.Fy = self._MF52.Fy(self.Fz, self.alpha, self.epsilon) * self.params.friction_scaling_y

    

File: julia/src/TireState.jl
# TireState.jl — included into module VogelSIM
# Depends on: MF52 (mf52_Fx, mf52_Fy)

mutable struct TireState
    mf52::MF52
    friction_scaling_x::Float64
    friction_scaling_y::Float64
    alpha::Float64    # slip angle (rad)
    kappa::Float64    # longitudinal slip
    epsilon::Float64  # inclination angle (rad)
    Fz::Float64       # normal load (N)
    Fx::Float64       # longitudinal force (N)
    Fy::Float64       # lateral force (N)
end

function TireState(mf52::MF52, fsx::Float64, fsy::Float64)
    return TireState(mf52, fsx, fsy, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0)
end

"""Peak longitudinal force by scanning kappa ∈ [0, 0.2] at 100 steps."""
function eval_Fx!(t::TireState)
    best = 0.0
    for sl in range(0.0, 0.2; length=100)
        val = mf52_Fx(t.mf52, t.Fz, sl, t.epsilon) * t.friction_scaling_x
        val > best && (best = val)
    end
    t.Fx = best
end

function eval_Fy!(t::TireState)
    t.Fy = mf52_Fy(t.mf52, t.Fz, t.alpha, t.epsilon) * t.friction_scaling_y
end
